pennylane/templates/state_preparations/mottonen.py

```diff
@@ -128,102 +134,76 @@ def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
+        array representing :math:`\alpha^{z,k}`
     """
-    alpha_z_k = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
+    alpha_z_k = np.zeros((2 ** (n - k),), dtype=np.float64)
 
-    for (i, _), om in omega.items():
-        i += 1
-        j = int(np.ceil(i * 2 ** (-k)))
+    for i in range(len(omega)):
+        j = int(np.ceil((i + 1) * 2 ** (-k)))
         s_condition = 2 ** (k - 1) * (2 * j - 1)
-        s_i = 1.0 if i > s_condition else -1.0
-        alpha_z_k[j - 1, 0] = alpha_z_k[j - 1, 0] + s_i * om / 2 ** (k - 1)
+        s_i = 1.0 if (i + 1) > s_condition else -1.0
+        alpha_z_k[j - 1] = alpha_z_k[j - 1] + s_i * omega[i] / 2 ** (k - 1)
 
     return alpha_z_k
 
 
 def _get_alpha_y(a, n, k):
-    r"""Computes the rotation angles alpha for the Y rotations.
+    r"""Computes the rotation angles for the Y rotations applied to the k'th qubit.
+
+    The angles are related to a as:
+
+    .. math:: \alpha^{y,k}_j = 2 \arcsin \sqrt{ \frac{ \sum_{l=1}^{2^{k-1}} a_{(2j-1)2^{k-1} +l}^2  }{ \sum_{l=1}^{2^{k}} a_{(j-1)2^{k} +l}^2  } }
 
     Args:
-        omega (float): phase of the input
+        a (float): absolute values of the input
         n (int): total number of qubits
         k (int): index of current qubit
 
     Returns:
-        scipy.sparse.dok_matrix[np.float64]: a sparse vector representing :math:`\alpha^y_k`
+        array representing :math:`\alpha^{y,k}`
     """
 
-    alpha = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-
-    numerator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-    denominator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
+    numerator = np.zeros((2 ** (n - k),), dtype=np.float64)
+    denominator = np.zeros((2 ** (n - k),), dtype=np.float64)
+    alpha = np.zeros((2 ** (n - k),), dtype=np.float64)
 
-    for (i, _), e in a.items():
-        j = int(math.ceil((i + 1) / 2 ** k))
-        l = (i + 1) - (2 * j - 1) * 2 ** (k - 1)
+    # compute all numerators/denominators at once for efficiency
+    for i in range(len(a)):
+        j_ = int(math.ceil((i + 1) / 2 ** k))
+        l = (i + 1) - (2 * j_ - 1) * 2 ** (k - 1)
         is_part_numerator = 1 <= l <= 2 ** (k - 1)
 
         if is_part_numerator:
-            numerator[j - 1, 0] += e * e
-        denominator[j - 1, 0] += e * e
+            numerator[j_ - 1] += a[i] * a[i]
+        denominator[j_ - 1] += a[i] * a[i]
 
-    for (j, _), e in numerator.items():
-        numerator[j, 0] = math.sqrt(e)
-    for (j, _), e in denominator.items():
-        denominator[j, 0] = 1 / math.sqrt(e)
+    for i in range(len(alpha)):
+        if denominator[i] != 0.0:
+            alpha[i] = numerator[i] / denominator[i]
 
-    pre_alpha = numerator.multiply(denominator)  # type: sparse.csr_matrix
-    for (j, _), e in pre_alpha.todok().items():
-        alpha[j, 0] = 2 * np.arcsin(e)
-
-    return alpha
+    return 2 * np.arcsin(np.sqrt(alpha))
```

Comment:
was tricky to follow the logic from old version to new version, but it seems correct :+1: